framework/filters/AccessRule.php

```diff
@@ -156,8 +171,13 @@ protected function matchRole($user)
                 if (!$user->getIsGuest()) {
                     return true;
                 }
-            } elseif ($user->can($role)) {
-                return true;
+            } else {
+                $params = empty($this->ruleParamsCallback) ? [] :
```

Comment:
I'd prefer a complete redesign of the core code than to add custom pieces to serve certain usecases. When done properly in an OO way, we can have code serving many more usecases (without ending up with fragmented code)